osf/management/commands/migrate_pagecounter_data.py

```diff
@@ -0,0 +1,120 @@
+import datetime
+import logging
+
+from django.core.management.base import BaseCommand
+from django.db import connection
+
+from framework import sentry
+
+logger = logging.getLogger(__name__)
+
+
+LIMIT_CLAUSE = ' LIMIT %s);'
+NO_LIMIT_CLAUSE = ');'
+
+REVERSE_SQL_BASE = '''
+UPDATE osf_pagecounter PC
+SET
+    resource_id = NULL,
+    file_id = NULL,
+    version = NULL
+WHERE PC.id in (
+    select PC.id from osf_pagecounter PC
+        left outer join osf_guid Guid on Guid._id = split_part(PC._id, ':', 2)
+        left outer join osf_basefilenode File on File._id = split_part(PC._id, ':', 3)
+    where
+          resource_id IS NOT NULL OR
+          file_id IS NOT NULL AND
+          Guid._id IS NOT NULL AND File._id IS NOT NULL
+'''
+REVERSE_SQL = '{} {}'.format(REVERSE_SQL_BASE, NO_LIMIT_CLAUSE)
+REVERSE_SQL_LIMITED = '{} {}'.format(REVERSE_SQL_BASE, LIMIT_CLAUSE)
+
+FORWARD_SQL_BASE = '''
+    UPDATE osf_pagecounter PC
+    SET
+        action = split_part(PC._id, ':', 1),
+        resource_id = Guid.id,
+        file_id = File.id,
+        version = NULLIF(split_part(PC._id, ':', 4), '')::int
+    FROM osf_guid Guid, osf_basefilenode File
+        WHERE
+              Guid._id = split_part(PC._id, ':', 2) AND
+              File._id = split_part(PC._id, ':', 3) AND
+              PC.id in (
+                  select PC.id from osf_pagecounter PC
+                      left outer join osf_guid Guid on Guid._id = split_part(PC._id, ':', 2)
+                      left outer join osf_basefilenode File on File._id = split_part(PC._id, ':', 3)
+                  where (PC.resource_id is NULL or PC.file_id IS NULL) AND
+                        Guid._id IS NOT NULL AND File._id IS NOT NULL
+'''
+FORWARD_SQL = '{} {}'.format(FORWARD_SQL_BASE, NO_LIMIT_CLAUSE)
+FORWARD_SQL_LIMITED = '{} {}'.format(FORWARD_SQL_BASE, LIMIT_CLAUSE)
+
+COUNT_SQL = '''
+select count(PC.id)
+    from osf_pagecounter as PC
+    left join osf_guid Guid on Guid._id = split_part(PC._id, ':', 2)
+    left join osf_basefilenode File on File._id = split_part(PC._id, ':', 3)
+where (PC.resource_id is NULL or PC.file_id IS NULL) AND
+      Guid._id IS NOT NULL AND File._id IS NOT NULL;
+'''
+
+
+class Command(BaseCommand):
+    help = '''Does the work of the pagecounter migration so that it can be done incrementally when convenient.
+    You will either need to set the page_size large enough to get all of the records, or you will need to run the
+    script multiple times until it tells you that it is done.'''
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            '--dry_run',
+            type=bool,
+            default=False,
+            help='Run queries but do not write files',
+        )
+        parser.add_argument(
+            '--page_size',
```

Comment:
Variable name is confusing to me.  I see that only one page of records is run, so if my page_size is 100, at most 100 records will be changed.  But the naming makes me think that it's going to be done in batches, though, 100 at a time until it's done.

Reply:
Part of that is leftover from when it was done in batches, and I had a `--sample-only` flag that would only do a single page. It still _technically_ works that way, just without the automatic going to the next page. I could call it `rows` or `rows_to_process`?

Reply:
I like either one!